RecoLocalCalo/EcalRecProducers/plugins/EcalPh2DigiToGPUProducer.cc

```diff
@@ -0,0 +1,135 @@
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  void acquire(edm::Event const& event, edm::EventSetup const& setup, edm::WaitingTaskWithArenaHolder holder) override;
+  void produce(edm::Event& evt, edm::EventSetup const& setup) override;
+
+private:
+  const edm::EDGetTokenT<EBDigiCollectionPh2> ebDigiCollectionToken_;
+  const edm::EDPutTokenT<cms::cuda::Product<ecal::DigisCollection<calo::common::DevStoragePolicy>>>
+      digisCollectionTokenEB_;
+  uint32_t neb_;
+
+  ecal::DigisCollection<::calo::common::DevStoragePolicy> ebdigis_;
+
+  cms::cuda::ContextState cudaState_;
+};
+
+void EcalPh2DigiToGPUProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("BarrelDigis", edm::InputTag("simEcalUnsuppressedDigis", ""));
+  desc.add<std::string>("digisLabelEB", "ebDigis");
+
+  descriptions.addWithDefaultLabel(desc);
+}
+
+EcalPh2DigiToGPUProducer::EcalPh2DigiToGPUProducer(const edm::ParameterSet& ps)
+    : ebDigiCollectionToken_(consumes<EBDigiCollectionPh2>(ps.getParameter<edm::InputTag>("BarrelDigis"))),
+      digisCollectionTokenEB_(produces<cms::cuda::Product<ecal::DigisCollection<calo::common::DevStoragePolicy>>>(
+          ps.getParameter<std::string>("digisLabelEB"))) {}
+
+void EcalPh2DigiToGPUProducer::acquire(edm::Event const& event,
+                                       edm::EventSetup const& setup,
+                                       edm::WaitingTaskWithArenaHolder holder) {
+  cms::cuda::ScopedContextAcquire ctx{event.streamID(), std::move(holder), cudaState_};
+
+  //input data from event
+  const EBDigiCollectionPh2* pdigis = &event.get(ebDigiCollectionToken_);
+
+  neb_ = pdigis->size();
+
+  //allocate device pointers for output
+  ebdigis_.ids = cms::cuda::make_device_unique<uint32_t[]>(neb_, ctx.stream());
+  ebdigis_.data = cms::cuda::make_device_unique<uint16_t[]>(neb_ * EcalDataFrame_Ph2::MAXSAMPLES, ctx.stream());
+
+  //allocate host vectors for holding product data and id vectors
+
+  std::vector<uint32_t, cms::cuda::HostAllocator<uint32_t>> idsebtmp;
+  std::vector<uint16_t, cms::cuda::HostAllocator<uint16_t>> dataebtmp;
+
+  //resize vectors to number of digis * size needed per digi
+
+  idsebtmp.resize(neb_);
+  dataebtmp.resize(neb_ * EcalDataFrame_Ph2::MAXSAMPLES);
+
+  //iterate over digis
+  uint32_t i = 0;
+  for (auto itdg = pdigis->begin(); itdg != pdigis->end(); ++itdg) {
+    //extract digi input
+    EcalDataFrame_Ph2 dataFrame(*itdg);
+
+    int nSamples = dataFrame.size();
+    //assign id to output vector
+    idsebtmp.data()[i] = dataFrame.id();
+
+    //iterate over sample in digi
+    for (int sample = 0; sample < nSamples; ++sample) {
+      //get samples from input digi
+      EcalLiteDTUSample thisSample = dataFrame[sample];
+      //assign adc data to output
+      dataebtmp.data()[i * nSamples + sample] = thisSample.raw();
+    }
+    i++;
```

Comment:
```suggestion
  for (const auto& pdigi : pdigis) {
    const int nSamples = pdigi.size();
    //assign id to output vector
    idsebtmp.data()[i] = pdigi.id();

    //iterate over sample in digi
    for (int sample = 0; sample < nSamples; ++sample) {
      //get samples from input digi
      EcalLiteDTUSample thisSample = pdigi[sample];
      //assign adc data to output
      dataebtmp.data()[i * nSamples + sample] = thisSample.raw();
    }
    ++i;
```
Better use a range based loop if possible.